rfd/0169-auto-updates-linux-agents.md

```diff
@@ -0,0 +1,364 @@
+  active_version: 15.1.1
+```
+
+backup.yaml:
+```
+version: v1
+kind: config_backup
+spec:
+  # proxy address from the backup
+  proxy: mytenant.teleport.sh
+  # version from the backup
+  version: 15.1.0
+  # time the backup was created
+  creation_time: 2020-12-09T16:09:53+00:00
+```
+
+### Runtime
+
+The agent-updater will run as a periodically executing systemd service which runs every 10 minutes.
+The systemd service will run:
+```shell
+$ teleport-updater update
+```
+
+After it is installed, the `update` subcommand will no-op when executed until configured with the `teleport-updater` command:
+```shell
+$ teleport-updater enable --proxy mytenant.teleport.sh
+```
+
+If the proxy address is not provided with `--proxy`, the current proxy address from `teleport.yaml` is used.
+
+On servers without Teleport installed already, the `enable` subcommand will change the behavior of `teleport-update update` to update teleport and restart the existing agent, if running.
+It will also run update teleport immediately, to ensure that subsequent executions succeed.
+
+The `enable` subcommand will:
+1. Configure `updates.yaml` with the current proxy address and set `enabled` to true.
+2. Query the `/v1/webapi/ping` endpoint.
+3. If the current updater-managed version of Teleport is the latest, and teleport package is not installed, quit.
+4. If the current updater-managed version of Teleport is the latest, but the teleport package is installed, jump to (12).
+5. Download the desired Teleport tarball specified by `agent_version` and `server_edition`.
+6. Verify the checksum.
+7. Extract the tarball to `/var/lib/teleport/versions/VERSION`.
+8. Replace any existing binaries or symlinks with symlinks to the current version.
+9. Backup /var/lib/teleport into `/var/lib/teleport/versions/OLD-VERSION/backup/teleport`
+10. Restart the agent if the systemd service is already enabled.
+11. Set `active_version` in `updates.yaml` if successful or not enabled.
+12. Replace the symlink/binary and `/var/lib/teleport` and quit (exit 1) if unsuccessful.
+13. Remove any `teleport` package if installed.
+14. Verify the symlinks to the active version still exists.
+15. Remove all stored versions of the agent except the current version and last working version.
+
+The `disable` subcommand will:
+1. Configure `updates.yaml` to set `enabled` to false.
+
+When `update` subcommand is otherwise executed, it will:
+1. Check `updates.yaml`, and quit (exit 0) if `enabled` is false, or quit (exit 1) if `enabled` is true and no proxy address is set.
+2. Query the `/v1/webapi/ping` endpoint.
+3. Check if the current time is after the time advertised in `agent_update_after`, and that `agent_auto_updates` is true.
+4. If the current version of Teleport is the latest, quit.
+5. Wait `random(0, agent_update_jitter_seconds)` seconds.
+6. Download the desired Teleport tarball specified by `agent_version` and `server_edition`.
```

Comment:
nit: before downloading a disk space should be reserved to make sure that enough space is available.